src/EntityComponentManager.cc

```diff
@@ -206,6 +209,79 @@ Entity EntityComponentManagerPrivate::CreateEntityImplementation(Entity _entity)
   return _entity;
 }
 
+/////////////////////////////////////////////////
+Entity EntityComponentManager::Clone(const Entity _entity, const Entity _parent,
+    const std::string &_name, bool _allowRename, bool _recursive)
+{
+  auto uniqueNameGenerated = false;
+
+  // Before cloning, we should make sure that:
+  //  1. The entity to be cloned exists
+  //  2. We can generate a unique name for the cloned entity
+  if (!this->HasEntity(_entity))
+  {
+    ignerr << "Requested to clone entity [" << _entity
+      << "], but this entity does not exist." << std::endl;
+    return kNullEntity;
+  }
+  else if (!_name.empty() && !_allowRename)
+  {
+    if (kNullEntity != this->EntityByComponents(components::Name(_name)))
+    {
+      ignerr << "Requested to clone entity [" << _entity
+        << "] with a name of [" << _name << "], but another entity already "
+        << "has this name." << std::endl;
+      return kNullEntity;
+    }
+    uniqueNameGenerated = true;
+  }
+
+  auto clonedEntity = this->CreateEntity();
+
+  if (_parent != kNullEntity)
+    this->CreateComponent(clonedEntity, components::ParentEntity(_parent));
```

Comment:
also call `this->SetParentEntity(clonedEntity, _parent);`?

Reply:
Yeah, thanks for catching this :+1: I added it in https://github.com/ignitionrobotics/ign-gazebo/pull/959/commits/8b0e17c2887028a8662aea087d2905a31679ecae